feat(ext/node): support path-scoped FFI for SQLite extension loading

charles-dyfis-net opened the pull request:
Previously, using `allowExtension: true` or calling `loadExtension()` required unrestricted `--allow-ffi` permission. This made it impossible to sandbox code that needs to load only specific, pre-approved SQLite extensions.

This change allows scoped FFI permissions:
- `allowExtension: true` now requires partial FFI permission (any scope)
- `loadExtension(path)` requires FFI permission covering that specific path

Example: `--allow-ffi=/path/to/extension.so` now permits loading only that extension, rather than granting unrestricted FFI access.

Fixes: https://github.com/denoland/deno/issues/31426

charles-dyfis-net committed:
feat(ext/node): support path-scoped FFI for SQLite extension loading

Previously, using `allowExtension: true` or calling `loadExtension()` required unrestricted `--allow-ffi` permission. This made it impossible to sandbox code that needs to load only specific, pre-approved SQLite extensions.

This change allows scoped FFI permissions:
- `allowExtension: true` no longer runs an up-front / connection-time check (the `partial` check function did not return true as expected in practice when only scoped FFI permissions exist)
- `loadExtension(path)` requires FFI permission covering that specific path

Example: `--allow-ffi=/path/to/extension.so` now permits loading only that extension, rather than granting unrestricted FFI access.

Note that this now universally disables the SQL `load_extension()` function, whether or not FFI is globally enabled.

Fixes: https://github.com/denoland/deno/issues/31426

charles-dyfis-net commented:
In attempting to address automated review feedback about docstrings, I appear to have broken the build -- it looks like `#[op2]` does not support `///` doc comments inside the impl block. Reverted those additions, while keeping the doc comment on `open_db`.

littledivy reviewed (commented):
AFAIU this does not prevent loading extension from SQL like so:
```sql
 SELECT load_extension("a.dll");
```
and that was the reason sqlite required unscoped ffi perms.

charles-dyfis-net commented:
> AFAIU this does not prevent loading extension from SQL like so:
> 
> ```sql
>  SELECT load_extension("a.dll");
> ```
> 
> and that was the reason sqlite required unscoped ffi perms.

Ah! Thank you for that clarification.

As I read https://sqlite.org/c3ref/c_dbconfig_defensive.html#sqlitedbconfigenableloadextension, it should be possible to enable only the C call and not the SQL -- I'll go back and extend the tests to look for that specific state, and then see about determining how to get into it.

charles-dyfis-net commented:
On updating the tests to distinguish between a normal pass and a test that was short-circuited due to the extension not being compiled, it looks like the extension tests were always being skipped because they were attempting to load the extension from the wrong location, triggering the early `return` path. When fixing that bug, at least on Darwin, I get a segfault akin to the following (and the exact same thing loading the extension from a separate `sqlite3` executable):

```
* thread #11, name = 'tokio-runtime-worker', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x0000000000000000
    frame #1: 0x000000013797ee9c libtest_sqlite_extension.dylib`mallocWithAlarm(n=40, pp=0x000000017095cff0) at sqlite3.c:30985:11
    frame #2: 0x0000000137978ee0 libtest_sqlite_extension.dylib`sqlite3Malloc(n=40) at sqlite3.c:31051:5
    frame #3: 0x0000000137982fa8 libtest_sqlite_extension.dylib`sqlite3HashInsert(pH=0x000000014b905a68, pKey="test_func", data=0x0000000140027f30) at sqlite3.c:37637:25
    frame #4: 0x000000013797ec8c libtest_sqlite_extension.dylib`sqlite3FindFunction(db=0x000000014b905800, zName="test_func", nArg=1, enc='\x01', createFlag='\x01') at sqlite3.c:129243:24
    frame #5: 0x000000013797e900 libtest_sqlite_extension.dylib`sqlite3CreateFunc(db=0x000000014b905800, zFunctionName="test_func", nArg=1, enc=1, pUserData=0x0000000000000000, xSFunc=(libtest_sqlite_extension.dylib`test_sqlite_extension::test_func::h6cec236b81aa5076 at lib.rs:12), xStep=0x0000000000000000, xFinal=0x0000000000000000, xValue=0x0000000000000000, xInverse=0x0000000000000000, pDestructor=0x0000000000000000) at sqlite3.c:184621:7
    frame #6: 0x000000013797e488 libtest_sqlite_extension.dylib`createFunctionApi(db=0x000000014b905800, zFunc="test_func", nArg=1, enc=2049, p=0x0000000000000000, xSFunc=(libtest_sqlite_extension.dylib`test_sqlite_extension::test_func::h6cec236b81aa5076 at lib.rs:12), xStep=0x0000000000000000, xFinal=0x0000000000000000, xValue=0x0000000000000000, xInverse=0x0000000000000000, xDestroy=0x0000000000000000) at sqlite3.c:184687:8
    frame #7: 0x000000013797e594 libtest_sqlite_extension.dylib`sqlite3_create_function_v2(db=0x000000014b905800, zFunc="test_func", nArg=1, enc=2049, p=0x0000000000000000, xSFunc=(libtest_sqlite_extension.dylib`test_sqlite_extension::test_func::h6cec236b81aa5076 at lib.rs:12), xStep=0x0000000000000000, xFinal=0x0000000000000000, xDestroy=0x0000000000000000) at sqlite3.c:184729:10
```

I suspect we'll want a separate PR to fix the extension used for test coverage before this one can be considered -- I'll get a separate ticket and PR together.

charles-dyfis-net commented:
Filed ticket #31454 for the issues preventing this from being correctly tested.

charles-dyfis-net commented:
About to push an updated version of this PR stacked on top of #31455. The expectation is that it will remain in draft until the latter (or something equivalent) is merged.

charles-dyfis-net commented:
(Most recent push is a rebase on top of current main to get CI fixes for macos-x86_64)

charles-dyfis-net commented:
@littledivy -- now that we're out of the holidays, where do we stand on this? I believe I've addressed your concerns by disabling the SQL interface to extension loading, but haven't yet heard whether that's actually the case.

bartlomieju committed:
Merge remote-tracking branch 'origin/main' into issue-31426

Resolved conflicts in ext/node_sqlite/database.rs (file moved from
ext/node/ops/sqlite/database.rs on main): kept the deferred, path-scoped
FFI permission check from this PR and main's new defensive mode and
initial limits handling.

bartlomieju commented:
I pushed a merge with main to resolve conflicts (the implementation moved to `ext/node_sqlite/database.rs` in #32142). The PR's logic is unchanged; it now coexists with the `defensive` mode and initial-limits handling added on main. I also fixed the stale doc comment on `load_extension` (no FFI permission is required at construction time, only at `loadExtension()` call time).

Regarding the review that requested changes: it misquotes the SQLite documentation. The actual text for [`SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION`](https://www.sqlite.org/c3ref/c_dbconfig_defensive.html#sqlitedbconfigenableloadextension) says:

> When the first argument to this interface is 1, then **only the C-API is enabled and the SQL function remains disabled**. If the first argument to this interface is 0, then both the C-API and the SQL function are disabled.

So this PR's approach is correct: `sqlite3_db_config(ENABLE_LOAD_EXTENSION, 1)` enables only `sqlite3_load_extension()`, which is gated by the path-scoped FFI check, while SQL `load_extension()` stays disabled. The test added in this PR ("SQL load_extension() is disabled even with allowExtension: true") verifies exactly that and passes on all platforms. This also matches Node.js: `node_sqlite.cc` uses `sqlite3_db_config(SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION, 1)` for `allowExtension` and never calls `sqlite3_enable_load_extension()`. @littledivy's original concern about the SQL-level bypass is therefore addressed; dismissing the stale review.

bartlomieju reviewed (commented):
Thanks for this — the scoped-FFI approach is a real sandboxing improvement, and the security model holds up: enabling `SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION` at open time only enables the C-API (the SQL `load_extension()` function stays disabled), and the sole load path now goes through the per-path FFI check. The dedicated test pinning that SQL-function-disabled behavior is a nice touch.

A few things worth addressing:

**1. The resolved path from the permission check is discarded (main one).** `check_ffi_partial_with_path` returns the canonicalized/resolved path, but here the return value is dropped and the raw `path` string is passed to `sqlite3_load_extension`. `ext/ffi/dlfcn.rs` uses the *returned* resolved path for the actual `dlopen`. Discarding it means the path that's checked and the path that's loaded can disagree under symlinks. Suggest using the returned path:

```rust
let path = state
  .borrow_mut::<PermissionsContainer>()
  .check_ffi_partial_with_path(Cow::Borrowed(Path::new(path)))?;
```

then build the `CString` from the resolved `path`.

**2. SQLite suffix expansion (minor).** `sqlite3_load_extension` may append a platform suffix (`.so`/`.dylib`/`.dll`) when the path lacks one, so granting `--allow-ffi=/x/ext` while SQLite actually loads `/x/ext.so` is a small check/load mismatch. The same theoretical gap exists for `dlopen`, so not a blocker, but a short comment noting that the granted scope should include the suffix would help future readers.

**3. Behavior change worth a release note.** `new DatabaseSync(":memory:", { allowExtension: true })` with no FFI permission now succeeds (it threw `NotCapable` before); the error surfaces only at `loadExtension()` time. This is the intended change and the test was updated accordingly, but it's user-visible and worth calling out in the changelog.

The `debug_assert_eq!(result, SQLITE_OK)` simplification is correct since the non-OK path already returns `Err` inside the `unsafe` block, and adding `--allow-run` for the subprocess-based scoped-FFI tests is reasonable. Test coverage is solid overall.

bartlomieju committed:
Merge remote-tracking branch 'origin/main' into issue-31426

# Conflicts:
#	ext/node_sqlite/database.rs

bartlomieju committed:
chore: keep node_compat suite pointer in sync with main
